Add PIE support for Xen and KVM

mogasergiu opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): `x86_64`, `arm64`
 - Platform(s): `kvm`, `xen`
 - Application(s):


### Additional configuration

`CONFIG_OPTIMIZE_PIE`
`CONFIG_UKPLAT_MEMREGION_MAX_COUNT`

### Description of changes

Drop the `elf32-i386` linker flag through a python script that converts an ELF64's Header, Section Headers and Program Headers to their 32-bit equivalents.

Make the Unikernel position independent with the help of  `struct uk_reloc`, a new structure accompanied by helper symbol-generating macros and a script that analyzes the final Unikernel binary such symbols as well as the `.rela.dyn` section's entries to build a binary blob inside a new custom ELF section `.uk_reloc` made up of `struct uk_reloc` entries that an early self relocator can make use of.

### NOTE
This is part of a larger set of Pull Requests. To make things easier and to ensure that things build as expected, it is recommended that testing is done based on #912, which includes everything. The splitting has been done to ease the review process.

mogasergiu committed:
plat/kvm: Drop `elf32-i386` output target linker flag

Remove the `elf32-i386` linker flag that forces the final Unikernel
binary output as a 32-bit kernel.

Since `qemu` currently boots us through `multiboot`, a boot protocol
that does not support 64-bit kernels, add a new support script
to be run on the final image. This `python` script will convert the
Unikernel's `ELF64` headers to their `ELF32` equivalents, thus fooling
the loader into thinking it's a 32-bit kernel. Since most of the
addresses contained into the fields of the headers are assumed to fit into
at most 4 bytes, the script simply takes those values and directly places
them in their corresponding `ELF32` header fields without truncating.

The actual `ELF32` header will be prepended to the final binary together
with the equivalent `ELF32` Program Headers and the Section Headers will
be appended to the end of the binary so that we can keep the `Multiboot`
header in the first 8192, as the specification requires. Although the
specification does not require the Program Headers be in the first
8192 bytes, GRUB seems to want it to be like that.

The script will be called using a newly added `Makefile.rules` file meant
to contain definitions common to all platforms.

Since, at this time, the only supported boot protocol on x86 KVM QEMU is
`multiboot`, enable this script for x86 KVM QEMU Multiboot builds.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/xen/x86: Use the already defined `ENTRY` macro from `uk/asm.h`

Remove the in-file definition of the `ENTRY` macro and use the already
existing macro from `uk/asm.h`.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/xen/x86: Make the Unikernel position independent

In order to be position independent, `struct uk_reloc` and the self
relocator based off it are used. Thus, the bootstrapping code calls
`do_uk_reloc` as early as possible, by using the page-sized default
stack that the Hypervisor passes onto us, because the actual stack
that we use, is also used by the Hypervisor and any early world switch
may cause the guest to crash.

Furthermore, the obvious absolute symbol references have been replaced
with their instruction pointer relative, position independent,
equivalents. The `r9` register has been used as a scratch register in
this case since, after a brief code analysis, it's been found to not be
used anywhere else during bootstrap.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
build: Add configuration option to build the Unikernel as a static PIE

Add compiler and linker options to properly build the Unikernel as a
static PIE. For now, only KVM and Xen are supported.

Strip `*dyn*` related ELF sections from the final binary. This will cause
tools such as `readelf` or `objdump` to issue some warnings or errors
when using on the final image due to the stripped unnecessary sections,
but they will still work and, as a trade off, the image will be as little
as it can be.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
lin/ukreloc: Add `ur_pte` macro to create relocatable PTE's

In order to cope with our static page tables used by the bootstrap code
we need to also be able to relocate our page table entries if we want
to be position independent on platforms that makes use of such page
tables.

This macro makes use of the already existing `ur_data` macro to make the
script create the usual `struct uk_reloc`. However a small modification
to the script makes the script do an additional lookup for corresponding
`pte_attr` symbols that `ur_pte` creates, so that it knows to also add
the page table entry attributes to the final value that the self
relocator uses to properly resolve a relocation.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/kvm/arm: Make the Unikernel position independent

Replace the static page table entries with relocatable ones through
the `ur_pte` macro. Thus, we make sure that relocations take place
before enabling the MMU by making a call to `do_uk_reloc` self relocator
method right before jumping to `start_mmu`.

Furthermore, make `Config.uk` select `LIBUKRELOC`.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
lib/ukreloc: Add ld script for sections required for a relocatable image

Add a separate Linker Script containing all the required sections
to achieve positional independence and the ability to self relocate.

The `.uk_reloc`, `.dynsym` and `.rela.dyn` sections will be part of the
main data segment and the last two will be stripped in the end, without
causing a hole in memory, due to the fact that `.uk_reloc` already
contains the forced `struct uk_reloc` relocations and, after being
updated, will also contain the `struct uk_reloc` equivalents of
`.rela.dyn`'s entries, which will be of the same size, so the `.bss`
section will stay in place.

The `.dynamic` and `.dynstr` sections become irrelevant in the final image
so we will strip them as well. In order not to cause a memory hole, place
them after `.comment` section.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/kvm/x86: Make the Unikernel position independent

Replace all `mov`'s and data declarations (`.long`, `.quad` etc.) that
use absolute symbol references with the equivalent `ur_*` macro's.

One exception to this type of relocation is going to be the 16-bit
code, as there is no point in relocating it before SMP initialization
without even knowing the physical address in the first 1MiB where the
secondary cores are to start execution. So, implement dedicated `ur_*`
macro's whose only use-case is the 16-bit code to cope with its
existence. Furthermore, add an exception to these `start16` relocation
symbols in the script.

Since, on this platform, we make use of static page tables, these were
also made relocatable through the `ur_pte` macro.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
lib/ukreloc: Implement a `struct uk_reloc` based self relocator

Add `do_uk_reloc` a method that parses the `.uk_reloc` section and
applies the relocations accordingly. As arguments, this method receives
a base physical address `r_paddr` for relocations that employ the
`UKRELOC_FLAGS_PHYS_REL` flag and a base virtual address `r_vaddr`
for those that do not want to be relocated against a physical address,
but rather by a given virtual address (e.g. can be used with KASLR).

In ARM's case, the `-fPIC`/`-fPIE`, with the help of `adrp`,
makes the code reference the page-aligned address of the section
and the offset this symbol has in this section where the relocations
are to be applied to and thus the desired relocated value is
obtained. Thus, unlike `x86` that simply does a `%rip` relative
access, ARM ends up dereferencing that in-section offset when trying
to get the runtime value of `__BASE_ADDR` which creates a chicken-egg
problem: we apply the relocations based on the current base address,
but in order to find the current base address we need to relocate
the value that is placed at the address of where the value of the
base address is supposed to be.
We solve this with the help of the `get_rt_addr()` function, which
forces an `adrp`, `add :lo12:` assembly sequence, which leads to
achieving something similar to x86's `%rip` relative addressing.

A case that may explain the need for `r_paddr` and `r_vaddr` would be
a situation where the previous program loader would place us at a random
physical address (r_paddr) and we then want to apply our own virtual
mappings for KASLR (r_vaddr), over the already applied virtual mappings,
if any.

Furthermore, right before applying relocations, the initial, added
through `mkbootinfo.py`, memory region descriptors are also relocated.
This is done by subtracting the initial link time base address, that
has been statically resolved by the linker in `lt_baddr`, and adding
the runtime base address.
Note: `lt_baddr` will generate a relocation so it will be overwritten
by the relocation loop. We want to keep it in case someone may want
to run the relocator multiple times for the same image. So, instead
of slowing down the relocator by checking each relocation against
`lt_baddr`'s  address so that we do not overwrite it, simply back
it up in a temporary variable and restore it after relocations are
done.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
lib/ukreloc: Add inline function to apply a `uk_reloc` entry

Implement a simple inline function to properly apply, depending on the
size of the relocation, a `uk_reloc` entry.

`apply_uk_reloc` may cause `UBSAN` to issue false positives, in
the case of the `x86` architecture which leads to the Unikernel crashing
since the serial console is not initialized yet, as it is mostly intended
to be called within the early self relocator. So make sure that
`UBSAN` does not touch `apply_uk_reloc` because it makes sense, for
`x86`, to make unaligned accesses (especially in Kernel Space) in order
to resolve non-compliant relocations.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
lib: Introduce `libukreloc`

In order to start supporting positional independence we need to be able
to self relocate. Thus, define a new custom relocation related structure
`struct uk_reloc` that is meant to hold the offset in memory where the
relocation is to be applied, the offset of the value from the original
symbol value computed by the linker through the linker script's base
reference address, the size of the relocation and a flags field
respectively (for now we only have the `UKRELOC_FLAGS_PHYS_REL` flag
meant to indicate whether the relocation is to be applied based on a
physical address or not, which comes in handy in bootstrap code).

The final binary blob that the `mkukreloc.py` script builds is made of
a signature (UKRELOC_SIGNATURE), `struct uk_reloc` entries appended
to each other and ends with a zeroed out `struct uk_reloc` to act as a
sentinel. This binary blob will be placed in the binary through the
`.uk_reloc` ELF section, updated through `objcopy`. Therefore, in order
for this forced section update to not mess up linker script exported
symbols, this section must be placed towards the end of the binary.

The binary blob is obtained by parsing all of the `.rela.dyn` entries,
converting them into `struct uk_reloc` entries and by searching
through the debug image's symbols for `_uk_reloc_` symbols created with
the help of the architecture independent helper assembly macro:
`ur_data`. This macro has been added to aid in replacing the references
to absolute symbol values with position independent equivalents that,
instead of referencing this very symbol, it places a placeholder value
(UKRELOC_PLACEHOLDER) and generates a symbol to be parsed by the script
to build a uk_reloc entry. Furthermore, it increases the size of the
`.uk_reloc` section by one entry through the `ur_sec_updt` macro. Note
that `ur_sec_updt` also inserts a dummy relocation entry in `.uk_reloc`
in order to force the linker to not optimize away the symbols. A unique
`uk_reloc` symbol will be generated through the `ur_sym` macro so that
an `ur_*` macro can be used on the same symbol more than once.
Consider the following usage example:
`ur_data quad, symbol, 8, _phys`
This will result in the `nm` tool, whose output the script will parse,
generating such entry:
`0000000000100106 T symbol_uk_reloc_data8_phys`
For `mkukreloc.py` this means, with a linker script reference address
of 0x100000:
```
struct uk_reloc {
        __u64 r_mem_off = 0x106;  // 0x100106 - 0x100000
        __u64 r_addr;
        __u32 r_sz = 8;  // symbol_uk_reloc_data[8]_phys
        __u32 flags = UKRELOC_FLAGS_PHYS_REL;  // symbol_uk_reloc_data8[_phys]
} __packed;
```
In order for actual value of `symbol` to be found, since its name is at
the beginning of the generated symbol, the script will now look for such
entry:
`0000000000100078 T symbol`  // from [symbol]_uk_reloc_data8_phys
and thus, `0x78` becomes the value of `r_addr` and the
`struct uk_reloc` is completed and appended to the binary blob with the
others.

`mkukreloc.py` will be invoked through `build_uk_reloc` Makefile definition
and the "section.*lma.*adjusted to.*" type of `objcopy` warnings will be
ignored since they are intended and harmless. The script will also check
that the distance between `.uk_reloc` and `.bss` is enough to not cause
an adjustment or spill. In case it is not enough, then the script will
raise an Exception.

All of this functionality shall be contained within the `libukreloc`
library.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat: Add `_base_addr` symbol to represent the reference base address

Export a linker script symbol called `_base_addr` to represent the exact
reference value used by the linker to resolve absolute references to
symbols.

Also, similarly to the other exported symbols, create a corresponding
macro, to avoid having to explicitly cast it for every reference.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/kvm/x86: Make SMP init code resolve its own `start16` relocations

Before waking up the secondary cores, the SMP initialization code copies
the, now position independent through `uk_reloc`, 16-bit and 32-bit
bootstrapping code to a physical page, in the lower 1 Mib of physical
memory. Since, at this point of execution, the immediate values used by
this bootstrapping code take the form of `UK_RELOC_PLACEHOLDER`, they
will need to be properly resolved after being moved into lower memory.

Therefore, add a new locally defined `struct uk_reloc` array holding the
hardcoded corresponding entries of these relocations, adapted to
reference the desired relocation address. Use this array after memory
copying the bootstrapping code to lower memory to resolve its
corresponding `start16` relocations.

Encode these entries with the help of `start16` related macro's and
since these definitions are starting to visually occupy a lot of space,
move everything to `start16_helpers.h`, a separate header file.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat: Group common linker script discards into one common macro

Add a macro-definition to contain all of the common sections that are
to be discarded and edit the linker scripts accordingly.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
lib/ukreloc/arch/arm64: Add position independent `ldr` macro

Create a custom `ur_ldr` macro to replace the `ldr` instruction in
places where references to absolute symbol values are made.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
lib/ukreloc/arch/x86_64: Add `ur_mov` instruction macro

Due to the high number of references to absolute symbol values in the
16-bit and 32-bit bootstrap code, a new macro, `ur_mov`, was introduced.
The new macro creates a new type of `_uk_reloc_` symbol, with an `imm`
suffix, to show that it is placed right after a `mov` instruction whose
immediate value must be relocated/patched. Thus, update `mkukreloc.py`
accordingly.

Furthermore, although useless, for completeness's sake add a 64-bit
variant. This variant does a `movabs` with the placeholder to
guarantee that the size of the immediate is 8 bytes.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
lib/ukreloc/arch/x86_64: Add 32-bit assembly self relocator

Similarly to the 64-bit C code `do_uk_reloc` self relocator,
`do_uk_reloc32` is a macro that takes as argument a value equal or
different from 0 to specify whether a stack is supplied or not. If the
stack is not supplied, the macro can generate its own minimal scratch
stack. Furthermore, the base virtual address is expected in %esi:%edi
(lower 32 bits in %edi and higher 32 bits in %esi), as well as the
physical base address in the %edx register. Since we are in Protected
Mode, we can assume a 32-bit register is enough to hold this address.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu commented:
v2:
- increase commit granularity
- Place the *dyn* related sections into SECT_STRIP_FLAGS to automatically get
stripped
- Place every *dyn* and .uk_reloc sections in a separate ld script that is
only considered if CONFIG_OPTIMIZE_PIE is enabled
- fix checkpatch warnings/errors
- add a new UK_RELOC_INJECT_COUNT to guarantee that uk_reloc does not spill
into other sections + mkukreloc.py checks the spill is not happening
- check against relocations generated by sections beyond .comment section
- wrap uk_reloc and elf64_to_32 scripts into build functions: build_elf64_to_32
and build_ukreloc
- remove unneeded arch specific `st_curr_baddr` in favor of `__BASE_ADDR` macro
- removed no longer needed `volatile` keyword leftovers from debugging
- tested on different UEFI machines (loaded at random addresses) with UBSAN
and ~28k relocations

mogasergiu commented:
v3:
- remove unnecessary final patching in `mkukreloc.py` as it doubles the script's execution time unnecessarily and makes the wrong assumption that offset in file == offset in memory, thus ending up causing wrong relocations in some unikernel images that do not happen to have the same layout by coincidence.
- rebase

mogasergiu committed:
support/scripts/mkbootinfo.py: Ensure Program Headers are sorted

Due to the placements of the `tls` and `tls_load` Program Headers,
The last two memory regions inserted in bootinfo structures will
always end up unordered. Thus, make sure they are ordered by their
address first, before building the memory regions and inserting
them.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/kvm/x86: Move `x86_start16_addr` out of `.data.boot`

Since `.data.boot` is meant mainly for boot specific data such as
a Multiboot header, move this unrelated `x86_start16_addr` out of
this section. The reason we do this is because this might cause
conflict with actual boot related data (e.g. `x86_start16_addr` be
placed at the beginning of the binary instead of the actual boot
related header data).

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

mogasergiu committed:
plat/kvm: Compute `ukplat_bootinfo` structure on the final binary

Since it is the final binary that is loaded and executed, make sure
that `build_bootinfo` is called on the final stripped binary, instead
of the debug image, whose segments' sizes may vary after stripping.

Signed-off-by: Sergiu Moga <sergiu.moga@protonmail.com>

dragosp27 reviewed (approved):
Great job!

Reviewed-by: Dragos Petre <dragos.petre27@gmail.com>

SerbanSo commented:
Looking good

Reviewed-by: Serban Sorohan serban.sorohan@gmail.com

jschlumpp commented on support/scripts/elf64_to_32.py:
```diff
@@ -0,0 +1,207 @@
+    elf32_phdrs = elf64.get_elf32_phdrs()
+    elf32_shdrs = elf64.get_elf32_shdrs()
+
+    with open(opt.elf64_file, 'r+b') as elf64_file:
```
Is there a reason to open it also as writeable?

jschlumpp commented on support/scripts/elf64_to_32.py:
```diff
@@ -0,0 +1,207 @@
+        elf64_buf = elf64_file.read()
+
+    # Create the empty file
+    with open(opt.output, 'w') as elf32_file:
```
Is there are reason to create the file first before writing?

jschlumpp commented on support/scripts/elf64_to_32.py:
```diff
@@ -0,0 +1,207 @@
+    with open(opt.output, 'w') as elf32_file:
+        pass
+
+    with open(opt.output, 'r+b') as elf32_file:
```
The file doesn't seem to be read so opening it for writing should be enough.

mogasergiu replied:
Hi Marco! Thank you a lot for the input :smiley:! Regarding your `elf64_to_32.py` comments, 4 months ago when I was writing the first versions of this script I was having two separate binaries: first I would just read the initial binary and then create the ELF32 equivalent. Unlike the current version that just prepends the ELF32 Headers. Indeed, the permissions wouldn't have made sense anyway even then, I think I was battling some permission issues at the time although I do not remember. I agree with you, I think I should have just done what I already do for [mkefi.py](https://github.com/unikraft/unikraft/pull/909/commits/e4a83a02b25cfac86414e1604710086dbd0f5e05#diff-7e2ddda8a9cccf09ea61481a4c03e5ffcff4bb1c0e671984681041a9d0e4acecR195) instead. So I am going to change these python lines accordingly :+1: .

jschlumpp commented on lib/ukreloc/include/uk/reloc.h:
```diff
@@ -0,0 +1,122 @@
+#define RELA_DYN_ENTRY_TYPE R_X86_64_64
+#elif defined(__ARM_64__)
+#define RELA_DYN_ENTRY_TYPE R_AARCH64_ABS64
+#endif
```
I would also add a `#else` branch that uses an `#error` pragma to fail the compilation with a message that indicates that a new case has to be added.

jschlumpp commented on support/scripts/mkukreloc.py:
I think the general direction was to avoid introducing further python scripts into the build process and instead implement these tools as small C programs using elfutils (libelf).

jschlumpp commented on lib/ukreloc/reloc.lds.S:
```diff
@@ -0,0 +1,62 @@
+		_ukreloc_start = .;
+		. += 4;
+		KEEP(*(.ukreloc))
+	} :data
```
You may want to use the [defines](https://github.com/unikraft/unikraft/blob/staging/plat/common/include/uk/plat/common/common.lds.h#L37-L45) for the segment annotations.

jschlumpp commented on plat/kvm/x86/lcpu_start.S:
It might be a good idea to add a note if relocations in this file change, that the corresponding `IMPORT_START16_UKRELOC_SYM` statements also have to be updated.

jschlumpp reviewed (commented):
Added a few notes

mogasergiu commented on lib/ukreloc/arch/x86_64/include/arch/reloc.h:
```diff
@@ -54,6 +54,159 @@
+ * @param have_stack Boolean value to tell whether the caller already has a
+ *                   stack available or not
+ */
+.macro do_ukreloc32	have_stack:req
```
I just noticed that this is going to break if, in the future, we make use of the MEMRT_KERNEL memory regions related to the ELF Segments as, unlike the C written relocator, it does not relocate the ELF Segment memory regions...

Unsure whether it is worth doing this iteration now or later when the time we will actually make use of the ELF Segment memory regions comes.

mogasergiu replied:
Introduced [`do_ukreloc_kmrds`](https://github.com/unikraft/unikraft/pull/772/commits/16ac1dbffc94f7570aab07c4f858893e6bf4cc54) for this.

mogasergiu commented:
Note for me. Rename `ukreloc` and related definitions to use `uk_reloc`

mogasergiu commented on plat/kvm/x86/multiboot.S:
```diff
@@ -77,10 +78,18 @@ ENTRY(_multiboot_entry)
 
-	movl	$lcpu_boot_startup_args, %edi	/* startup args for boot CPU */
+	/* Hardcoding for now I guess... */
+	movl    $0x00100000, %edi
```
These 3 lines I do not like, maybe some way to add an additional configuration option for the load address in the linker script, but while allowing three different versions (e.g. if a user wants to build KVM, Xen and linuxu images with the same configurations at the same time). So RFC.

razvand replied:
I would leave them as such for the time being. Let's have something working first and wait for the time when a configurable setup is required.

mogasergiu committed:
lib/ukreloc: Implement `do_uk_reloc_kmrds` to relocate Kernel segments

Pull this functionality out of `do_uk_reloc` so that other code has
access to the ability to relocate the Kernel ELF segments.

Since this must be used with caution and in very restrictive
conditions, a corresponding warning has been added to this function.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

razvanvirtan commented on lib/ukreloc/include/uk/reloc.h:
```diff
@@ -0,0 +1,122 @@
+#ifndef __UKRELOC_H__
+#define __UKRELOC_H__
+
+#define UKRELOC_FLAGS_PHYS_REL		(1 << 0)
```
Is there one extra-tab here? :thinking:

michpappas commented on plat/linuxu/arm64/entry64.S:
```diff
@@ -35,6 +35,13 @@
 
 .section .text
 
+/* For now, although inaccurate, set the
```
Although I understand what this comment refers to, it's only because I'm looking at it through the context of the PIE PR. Yet chances are that it's not clear to someone looking at this file out of scope. Perhaps you can provide some reasoning on both why this is added here, and why it doesn't matter in this platform. Same to the other variants where this comment applies.

michpappas commented on support/scripts/elf64_to_32.py:
```diff
@@ -0,0 +1,201 @@
+        _endian = '>' if self.endian == 'big' else '<'
+
+        # Check e_machine
+        if sum(self.elf64_ehdr[18:20]) == ELF_MACHINE['arm64']:
```
From my understanding this script is not used in `arm`/`arm64`, so perhaps this and the ELF magic for arm are not required?

michpappas commented on plat/kvm/Linker.uk:
```diff
@@ -67,6 +67,9 @@ $(KVM_IMAGE): $(KVM_IMAGE).dbg
 			{ grep -v "Empty loadable segment detected" || true; })
 	$(call build_bootinfo,$@)
+ifeq ($(ELF64_TO_32),y)
+	$(call build_elf64_to_32,$@)
```
Leaving this here as GitHub does not allow commenting on the commit messages. Since we now do support lxboot (to some extent), the commit message can be updated to use a more accurate description.

michpappas commented on lib/ukreloc/Config.uk:
```diff
@@ -0,0 +1,3 @@
+config LIBUKRELOC
+	bool "ukreloc: Self relocation and symbol based relocations"
+	default y
```
Why does `libukreloc` need to be enabled by default?

mogasergiu replied:
`libukreloc` needs to be enabled by default to avoid surrounding calls to `ur_*` macros with `#ifdef's`. The `ur_*` macro's, when `CONFIG_OPTIMIZE_PIE` is disables, revert back to using absolute addressing.

michpappas commented on lib/ukreloc/reloc.lds.S:
```diff
@@ -0,0 +1,63 @@
+	.ukreloc :
+	{
+		_ukreloc_start = .;
+		. += 4;
```
Ah, the good old magic 4... 🔮  Maybe use an ALIGN here?

mogasergiu replied:
That's supposed to be there to make room for the 4byte signature, but you're right I should have a locally defined macro for it. `ALIGN`? What for?

michpappas replied:
Okay, in that case a `#define` would do

michpappas commented on plat/common/arm/cpu_native.c:
```diff
@@ -56,10 +56,8 @@ void reset(void)
 	 * halt the CPU.
 	 */
-	if (!smccc_psci_call) {
-		uk_pr_crit("Couldn't reset system, HALT!\n");
```
So these messages are generally useful, and I know for fact that the pl011 and ns16550 drivers have no issue if the UART is not initialized, as they just return, so if the problem is caused by the x86 drivers a fix can be applied there instead?

If you do apply this change, also remember to update the commit message of the previous commit:
```
    `apply_ukreloc` may cause `UBSAN` to issue false positives, in
    the case of the `x86` architecture which leads to the Unikernel crashing
    since the serial console is not initialized yet, as it is mostly intended
    to be called within the early self relocator. So make sure that
    `UBSAN` does not touch `apply_ukreloc` because it makes sense, for
    `x86`, to make unaligned accesses (especially in Kernel Space) in order
    to resolve non-compliant relocations.
```
With the opportunity, I find the argumentation regarding UART a bit weak here. If unaligned accesses are normal here (which as discussed they are on x86 early boot code), then that should be enough reasoning regardless of the UART.

michpappas commented on lib/ukreloc/arch/arm64/include/arch/reloc.h:
The plat-rearch way is to create a hierarchy as `arch/<arch>/include/uk/asm/` and then in your `include/uk/reloc.h` you do
```
#include <asm/reloc.h>
```
To get this to work you need to update your `Makefile.uk` to also add the arch-specific path to the `CINCLUDES`.

And finally update any code to indluce `uk/reloc.h` instead of `arch/reloc.h`.

michpappas commented on arch/arm/arm64/ctx.S:
```diff
@@ -132,7 +132,8 @@ ENTRY(ukarch_ctx_switch)
 	 * its callee-saved registers in next switch to time.
 	 */
-	ldr x30, =restore_point
+	adrp x30, restore_point
```
Any reason `ur_ldr` is not used here?

mogasergiu replied:
IIRC the inclusion path for `arch/` assembly files do not take libraries into account. So I left it like this in the hopes that plat re-arch will fix this.

michpappas replied:
Yes that's true. Let's do that then 👍🏼

michpappas commented on lib/ukreloc/reloc.c:
```diff
@@ -79,18 +78,8 @@ void __used do_ukreloc(__paddr_t r_paddr, __vaddr_t r_vaddr)
-		mrdp->vbase += r_vaddr;
-	}
+	/* Relocate UKPLAT_MEMRT_KERNEL memory region descriptors */
+	do_ukreloc_kmrds(r_paddr, r_vaddr);
```
This breaks builds that don't enable `OPTIMIZE_PIE`, maybe make it conditional.

mogasergiu replied:
I will create a stub for it just like for `do_ukreloc`

michpappas replied:
I think the right way is to make it conditional, similarly to calls from other libraries. As discussed, in the future the `ur_` macros will become unconditionally available and the library won't be enabled by default, but that doesn't mean that all other definitions from this library will become unconditionally available too.

mogasergiu replied:
I do not like having `#ifdef`'s that could have been easily hidden away from the client code. I like having something similar to what linux does: if the CONFIG for this function is not enabled, then use a stub instead. If you place `#ifdef`'s throughout client code it's going to end up being an unreadable, unmaintainable mess. My opinion is that this kind of situations should be transparently handled by the server of the API's.

michpappas replied:
Linux is not compiled into individual libraries. In any case, using `#ifdef` is the way currently done, so any proposal to change this scheme shoud be discussed in a broader scope rather being applied ad-hoc.

mogasergiu replied:
> Linux is not compiled into individual libraries.

Fair, but the build process is largely the same. The burden of `#ifdef`-ing should not fall on the client code. IMO it would be better if it would be handled by the build system.

> In any case, using #ifdef is the way currently done,

And do you not agree with me that it looks like a mess? It's enough to look at `ukboot/boot.c`.

> any proposal to change this scheme shoud be discussed in a broader scope rather being applied ad-hoc.

This library is going to stay implicitly enabled for a long time. When that does not happen I think we should put the burden of `#ifdef`-ing on the build system, e.g. if the library is not enabled, include a dummy header with stubs from `include/uk`.

> And do you not agree with me that it looks like a mess

If you agree with me in this regard, why not begin with this library.

michpappas replied:
> This library is going to stay implicitly enabled for a long time

Actually not, the plan is to address this in the next release.

> If you agree with me in this regard, why not begin with this library.

Because it is not just my call, and it needs to be agreed by the rest of the maintainers.

michpappas replied:
> I will create a stub for it just like for `do_ukreloc`

Actually I was referring to calls to `do_ukreloc` and `do_ukreloc_mrds`, but accidentally placed my comment in the definition instead.

The same comment applies to `do_ukreloc`.

mogasergiu replied:
I know, I still do not agree.

michpappas commented on support/scripts/mkukreloc.py:
```diff
@@ -0,0 +1,325 @@
+# command's output...).
+# Luckily we do not need to deal with .symtab since `readelf -r` nicely shows
+# the corresponding st_value.
+R_RELATIVE_TYPE_EXP = r'R_(?:X86_64|AARCH64)_RELATIV[E]{0,1}'
```
Perhaps it would be useful to additionally check for unexpected relocation types.

mogasergiu replied:
That is done by this [check](https://github.com/unikraft/unikraft/pull/772/files#diff-1b6f1a9686178d25fb2b6e0f9ce8a803518660fed6a76c425c53a78df8e44acfR307). If the relocation is unexpected it will not be detected through the regex, leading to us finding less relocations than `.dynamic` reports.

michpappas replied:
Yes I did not expect that this is done by the regex of course. Didn't notice the [check](https://github.com/unikraft/unikraft/pull/772/files#diff-1b6f1a9686178d25fb2b6e0f9ce8a803518660fed6a76c425c53a78df8e44acfR307), thanks 👍🏼

razvand commented on Config.uk:
```diff
@@ -95,6 +95,13 @@ config OPTIMIZE_LTO
 		will increase overall building time but creates more efficient
 		Unikraft binaries.
 
+config OPTIMIZE_PIE
```
Why is this `OPTIMIZE_PIE` and not just `PIE`? It's not an optimization, is it?

mogasergiu replied:
I took @skuenzer's [suggestion](https://github.com/unikraft/unikraft/pull/239#discussion_r668293448) into consideration.

razvand commented on Makefile.uk:
```diff
@@ -88,15 +88,32 @@ M4FLAGS      += -DUK_VERSION=$(UK_VERSION).$(UK_SUBVERSION)
 # "--enable-default-pie" by default.
+ifeq ($(CONFIG_OPTIMIZE_PIE),y)
+COMPFLAGS-$(call gcc_version_ge,6,1)	+= -fPIE
+LDFLAGS-$(call gcc_version_ge,6,1)	+= -static-pie
```
From my understanding, the `static-pie` flag is used beginning with GCC 8: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=81498

mogasergiu replied:
I need this to be `-static-pie` and not just `-pie`, because otherwise you are risking having GOT-related relocations generated which I do not and will not handle just to support an older version of the compiler because it does not make any sense to have this kind of relocations present in a kernel.

EDIT: I am going to have a try with `-static -pie` though in the hopes that it will work the same :crossed_fingers:

mogasergiu replied:
:disappointed: `readelf: Error: no .dynamic section in the dynamic segment`. I would leave it with `-static-pie` for now... If the time comes that this would cause issues... I will open up an issue and look into it.

razvand replied:
My point was that you have to check for GCC >= 8 for this to work. If GCC < 8, you have to issue an error when `CONFIG_OPTIMIZE_PIE` is enabled.

razvand commented on Makefile.uk:
```diff
@@ -88,15 +88,32 @@ M4FLAGS      += -DUK_VERSION=$(UK_VERSION).$(UK_SUBVERSION)
+else
 COMPFLAGS-$(call gcc_version_ge,6,1)	+= -fno-PIC
 LDFLAGS-$(call gcc_version_ge,6,1)	+= -no-pie
+COMPFLAGS-$(call have_clang)	+= -fno-PIC
+LDFLAGS-$(call have_clang)	+= -no-pie
```
Why only for clang? These should apply to GCC as well, right? In case the `OPTIMIZE_PIE` option is not selected.

mogasergiu replied:
But they are, right above these lines, aren't they? :confused:

razvand replied:
I'm stupid. You're right.

razvand commented on lib/ukreloc/Config.uk:
```diff
@@ -0,0 +1,3 @@
+config LIBUKRELOC
+	bool "ukreloc: Self relocation and symbol based relocations"
```
```suggestion
	bool "ukreloc: Self relocation and symbol-based relocations"
```

razvand commented on lib/ukreloc/arch/arm64/include/arch/reloc.h:
```diff
@@ -0,0 +1,32 @@
+#ifdef __ASSEMBLY__
+
+/*
+ * Macro to replaced the non-relocation fiendly `ldr` instruction. The usage
```
```suggestion
 * Macro to replace the non-relocation-friendly `ldr` instruction. The usage
```

razvand commented on lib/ukreloc/arch/x86_64/include/arch/reloc.h:
```diff
@@ -0,0 +1,220 @@
+
+#ifdef __ASSEMBLY__
+
+/* Relocation friendly ur_mov to replace mov instructions incompatible with
```
```suggestion
/* Relocation-friendly ur_mov to replace mov instructions incompatible with
```

razvand commented on lib/ukreloc/arch/x86_64/include/arch/reloc.h:
````diff
@@ -0,0 +1,220 @@
+ * ```
+ * ur_mov	gdt64_ptr, %eax, 4
+ * ```
+ * The above will make mkukreloc.py process the symbol gdt64_ptr_ukeloc_imm4
````
```suggestion
 * The above will get mkukreloc.py to process the symbol gdt64_ptr_ukeloc_imm4
```

razvand commented on lib/ukreloc/arch/x86_64/include/arch/reloc.h:
```diff
@@ -0,0 +1,220 @@
+ * if there.
+ *
+ * NOTE: Unlike its C-variant, do_ukreloc(), this does not relocate
+ * the Kernel's static, build-time memory region descriptors!
```
```suggestion
 * the kernel's static, build-time memory region descriptors!
```

razvand reviewed (changes requested):
@mogasergiu, did a first quick pass. Will also test extensively and come back with another deeper pass.

razvand commented on lib/ukreloc/include/uk/reloc.h:
```diff
@@ -0,0 +1,217 @@
+#define __UKRELOC_H__
+
+#define UKRELOC_FLAGS_PHYS_REL		(1 << 0)
+#define UKRELOC_PLACEHOLDER		0xB00B
```
```suggestion
#define UKRELOC_PLACEHOLDER		0xB0B0
```

razvand commented on lib/ukreloc/include/uk/reloc.h:
```diff
@@ -0,0 +1,217 @@
+
+#define UKRELOC_FLAGS_PHYS_REL		(1 << 0)
+#define UKRELOC_PLACEHOLDER		0xB00B
+#define UKRELOC_SIGNATURE		0xBADB0111
```
```suggestion
#define UKRELOC_SIGNATURE		0x0BADB0B0
```

razvand commented on lib/ukreloc/include/uk/reloc.h:
```diff
@@ -0,0 +1,217 @@
+ * You may not use this file except in compliance with the License.
+ */
+
+#ifndef __UKRELOC_H__
+#define __UKRELOC_H__
```
```suggestion
#ifndef __UK_RELOC_H__
#define __UK_RELOC_H__
```

michpappas reviewed (changes requested):
@mogasergiu great work 🚀 Some minor review comments below.

razvand reviewed (changes requested):
For commit https://github.com/unikraft/unikraft/pull/772/commits/7d3053e9df912d9df0955f2da06143f4cf5490fc, use `program headers, the last ...` (aka lower case `p` and `h` and `t`).

razvand commented on plat/common/include/uk/plat/common/sections.h:
```diff
@@ -43,6 +43,9 @@
 /* _dtb: start of device tree */
 extern char _dtb[];
 
+/* _base_addr: base reference address as of link time */
```
I think this comment could be improved. If my understanding is correct, I would reword this to "This is the base address of the unikernel image. It is set at link time. It is used to compute other addresses, as reference to this.".

mogasergiu commented on support/scripts/elf64_to_32.py:
```diff
@@ -0,0 +1,201 @@
+        # Write equivalent ELF32 header
+        elf32.write(elf32_ehdr)
+
+        # Write the ELF32 equivalent Program Headers at the end of the binary
```
s/binary/ELF32 header

razvand reviewed (changes requested):
Be wary of the reference to `ukplat_memregion_list_coalesce()` which is only later defined in #848. Other than that, this is a great PR: commits, comments, everything looks great.

I'll do extensive testing after you integrate all changes.

mogasergiu committed:
plat/linuxu: Add dependency on `PIE` not being enabled

Building `linuxu` with the `PIE` build options results in errors
being issued as the way we currently implement PIE is not suited
for this platform yet.

Therefore, make sure that it is impossible to build `linuxu` as
long as `PIE` is enabled.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

michpappas reviewed (approved):
After offline discussion with @mogasergiu, @mschlumpp, @razvand there are still open questions, yet due to: 1. the extensive length of this PR, 2. the amount of testing required with every update, and 3. the time pressure to the upcoming release the consensus was to merge as is and provide a follow-up update after the maintainers team concludes on the remaining topics.

CC: @skuenzer 

Reviewed-by: Michalis Pappas <michalis@unikraft.io>

razvand reviewed (approved):
Reviewed-by: Razvan Deaconescu <razvand@unikraft.io>
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
